### PR TITLE
added exporter for pwdsphinx

### PR DESCRIPTION
**Pull request by stef**
This PR introduces support for pwdsphinx: https://www.ctrlc.hu/~stef/blog/posts/sphinx.html

SPHINX is an online "password store" that follows the KISS principle and has stronger than usual security guarantees. It is not possible to export passwords from SPHINX due to there not being any native support for listing all records belonging to a user. Thus thus this PR introduces only an exporter to SPHINX this causes some trouble as in https://github.com/stef/pass-import/blob/master/CONTRIBUTING.md#how-to-add-the-support-for-a-new-password-manager mandates:
> Add a file named tests/assets/db/mymanager[.csv,.xml,...]. No contribution will be accepted without this file. This file must contain the exported data from your manager.

Since there is not export functionality possible with SPHINX it is impossible to provide such a file. I hope despite this you accept this PR as it is not really a violation of the above mandate since such a file can only be demanded for importers not exporters. (huh it's kinda confusing what is an export and what an import, it all depends on the perspective i guess).

thanks a lot for your awesome project.

### CODE REVIEW

**Comment by stef** [comment]
hrmpf. CI obviously fails since i did not add pwdsphinx as a dep in travis/gitlab CI config. the problem is pwdsphinx depends on two libs which are not yet in debian (but are already packaged only need to be added to unstable).

i understand if the lack of deps is a blocking reason for acceptance of this PR.

i could provide the two libs online for download into the CI VMsif that would be acceptable?

**Comment by stef** [comment]
i managed to mock/fake the missing libs in travis-ci, however at the cost of moving to bionic from xenial and going from py3.6 to py3.7, i dunno if that is acceptable. i do hope it is. i have not touched the .gitlab-ci config, that if there is a gitlab hosting this somewhere will probably fail.

**Comment by roddhjav** [comment]
Hi, Thank a lot for this PR. I will merge it, just allow me a bit of time (pwdsphinx seems quite neat, I want to test it too).

Meanwhile, I have a few comments on your PR.
* Do not worries for the Travis conf. At some point I will move to Github Action anyway. The real CI is done on [Gitlab](https://gitlab.com/roddhjav/pass-import)
* As it is not a common format (like in XML, CSV...) you should remove `pass_import/formats/sphinx.py ` and put all the logic in `pass_import/managers/sphinx.py`.
* Looking at the test result, I see you only have 73% of code coverage. It indicate you might need to add more tests. Note you don't need 100%, but 73% is very low.

> (huh it's kinda confusing what is an export and what an import, it all depends on the perspective i guess).

You are right. In pass-import jargon an importer is a source password manager whereas an exported is a destination password manager. The jargon comes from the time pass was the only destination pm possible and therefore all other password manager were importers.

> Since there is not export functionality possible with SPHINX it is impossible to provide such a file.

No problem with the actual file (that more for CSV and similar exported file). But it seems possible to export the data from sphinx to another password manager simply because you can list and get the data stored in sphink. So this integration would be nice to add.

**Comment by stef** [comment]
> Hi, Thank a lot for this PR. I will merge it, just allow me a bit of time (pwdsphinx seems quite neat, I want to test it too).

thanks a lot!

> Meanwhile, I have a few comments on your PR.
> 
>     * As it is not a common format (like in XML, CSV...) you should remove `pass_import/formats/sphinx.py ` and put all the logic in `pass_import/managers/sphinx.py`.

ack. makes sense. will do so.

>     * Looking at the test result, I see you only have 73% of code coverage. It indicate you might need to add more tests. Note you don't need 100%, but 73% is very low.

right. will add more tests.

> > Since there is not export functionality possible with SPHINX it is impossible to provide such a file.
> 
> No problem with the actual file (that more for CSV and similar exported file). But it seems possible to export the data from sphinx to another password manager simply because you can list and get the data stored in sphinx. So this integration would be nice to add.

listing is not quite possible. you have to know the "hostnames" or "urls" for which you want to query the known users names. but the client itself does not handle any such lists of hostnames/urls/whatever. since that would mean the client has to sync this information to all other devices, and the whole idea is to not have to sync anything. so yes in theory if you give sphinx a hostname like "google" it will give you a list of usernames associated with you and google, which you then can query individually, but there is no complete list. although i also have to say that in the contribs there's some x11 integration with dmenu which caches hostnames, but that is independent of the client itself.

i'll update the PR ASAP with moving format to manager and adding more tests.

thanks for your comments.

**Comment by stef** [comment]
i updated the PR according to your comments.

File: pass_import/formats/sphinx.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                print(f"\t{id}")
+            self.reused[pwd].append(f"{user}@{host}")
+        else:
+            self.reused[pwd]=[f"{user}@{host}"]
+
+        try: sphinx.bin2pass.pass2bin(pwd, None)
+        except OverflowError:
+            raise PMError(f"Failed to import password for {user}@{host} to sphinx: too big.")
+
+        s = sphinx.connect()
```
[the remark this answers is not in the document]
Comment (stef) [reply pass_import/formats/sphinx.py]: currently every request must be a separate connection.

File: pass_import/formats/sphinx.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            self.weak+=1
+            print(f"Password for {user}@{host} is weak consider changing it asap with sphinx.")
+            #print(q)
+        if pwd in self.reused:
+            print(f"Password for {user}@{host} is reused also for:")
+            for id in self.reused[pwd]:
+                print(f"\t{id}")
+            self.reused[pwd].append(f"{user}@{host}")
+        else:
+            self.reused[pwd]=[f"{user}@{host}"]
```
[the remark this answers is not in the document]
Comment (stef) [reply pass_import/formats/sphinx.py]: agreed, where would this be better suited then?

File: pass_import/formats/sphinx.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+from pass_import.manager import PasswordExporter
+from pass_import.errors import PMError
+from pass_import.tools import getpassword
+
+class SPHINX(PasswordExporter):
+    """Base class for SPHINX based exporter.
+
+    This exporter requires pwdsphinx to run.
+    """
+    cap = Cap.EXPORT
```
[the remark this answers is not in the document]
Comment (stef) [reply pass_import/formats/sphinx.py]: it seems that this is an outdated version of the file.

[the remark this answers is not in the document]
Comment (stef) [reply pass_import/formats/sphinx.py]: this file doesn't exist anymore. i removed this from ./pass_import/managers/sphinx.py instead.

File: pass_import/formats/sphinx.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stef) [reply pass_import/formats/sphinx.py]: i removed it. but would love to see it pop up somewhere else in the code where it better fits. ;)

File: pass_import/formats/sphinx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (roddhjav) [reply pass_import/formats/sphinx.py]: Yes, I wrote this review a while ago but I forgot to submit it. So it is on a former version of the PR.

File: pass_import/formats/sphinx.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (roddhjav) [reply pass_import/formats/sphinx.py]: Right now, it is kind of already present (with more feature) in [pass-audit]((https://github.com/roddhjav/pass-audit0. See [`pass_audit/audit.py`](https://github.com/roddhjav/pass-audit/blob/master/pass_audit/audit.py) for the code.

File: pass_import/formats/sphinx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stef) [reply pass_import/formats/sphinx.py]: that makes sense.

File: pass_import/formats/sphinx.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stef) [reply pass_import/formats/sphinx.py]: TIL. this looks awesome. however with sphinx there's this problem of naturally not being able to enumerate all passwords and thus a conversion like this is the only moment when all passwords can be audited. is there a way to have pass-audit being used during a import?

**Comment by stef** [comment]
hmmm, i don't quite see why the build is failing here, i don't think it is related to my changes.

**Comment by roddhjav** [comment]
> is there a way to have pass-audit being used during a import?

Yes, I can call pass-audit from pass-import like

> i don't quite see why the build is failing here, i don't think it is related to my changes.

It is not related to your last changes but it is linked with your branch. The master branch still runs fine.

**Comment by stef** [comment]
> > is there a way to have pass-audit being used during a import?
> 
> Yes, I can call pass-audit from pass-import like

i searched in this repo for references for audit, but couldn't find any good explanation. how can audit be used during an import?

> > i don't quite see why the build is failing here, i don't think it is related to my changes.
> 
> It is not related to your last changes but it is linked with your branch. The master branch still runs fine.

weird.

**Comment by roddhjav** [comment]
A have now added the support pass password audit. See caf32e5b220f9bc52a9ea5bf276297829d6c6bbc

**Comment by stef** [comment]
that is so awesome! <3 <3 <3
thanks

**Comment by roddhjav** [comment]
Thanks. Merged!

**Comment by stef** [comment]
there's a new major version of pwdsphinx which is also backward incompatible, i'll send an update soon. in the mean time see https://sphinx.pm/ :) also we are now solid in debian an derivatives, apt install pwdsphinx, should work up to debian testing, and soon also stable (next stable release)

**Pull request merged**